File: list/Solution86.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution86 {
    public ListNode partition(ListNode head, int x) {
    	ListNode small = new ListNode();
        ListNode large = new ListNode();
        ListNode point1 = small;
        ListNode point2 = large;

        while(head != null){
            if(head.val >= x){
                point2.next = head;
                point2 = point2.next;
            }

            else{
                point1.next = head;
                point1 = point1.next;
            }
            head = head.next;
        }

        point1.next = large.next;
        point2.next = null;
        return small.next;
    }

    public static void main(String[] args){
    	Solution86 s = new Solution86();


        ListNode test2 = new ListNode(1, new ListNode(2, new ListNode(3)));
        s.partition(test2, 4);
    }
}

File: C++/Solution86.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<stack>
using namespace std;


 // Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution86 {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* sentinel = new ListNode();
        ListNode* small = new ListNode();
        ListNode* large = new ListNode();
        ListNode* ptr = head;
        ListNode* s = small;
        ListNode* l = large;
        while(ptr){
            if(ptr -> val < x){
                s -> next = new ListNode(ptr -> val);
                s = s -> next;
            }
            else{
                l -> next = new ListNode(ptr -> val);
                l = l -> next;
            }
            ptr = ptr -> next;
        }
        if(small -> next){
            sentinel -> next = small -> next;
            s -> next = large -> next;
            return sentinel -> next;
        }
        return head;
    }
};